server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerSenderRateLimit.scala

```diff
@@ -0,0 +1,211 @@
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.mailets
+
+import java.time.Duration
+import java.time.temporal.ChronoUnit
+
+import com.google.common.annotations.VisibleForTesting
+import eu.timepit.refined.auto._
+import javax.inject.Inject
+import org.apache.james.core.MailAddress
+import org.apache.james.rate.limiter.api.Quantity.Quantity
+import org.apache.james.rate.limiter.api.{AcceptableRate, Quantity, RateExceeded, RateLimiter, RateLimiterFactoryProvider, RateLimitingKey, RateLimitingResult, Rule, Rules}
+import org.apache.james.util.DurationParser
+import org.apache.mailet.Mail
+import org.apache.mailet.base.GenericMailet
+import org.reactivestreams.Publisher
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+case class KeyPrefix(value: String)
+
+sealed trait EntityType {
+  def asString(): String
+
+  def extractQuantity(mail: Mail): Quantity
+}
+
+case object Count extends EntityType {
+  override def asString(): String = "count"
+
+  override def extractQuantity(mail: Mail): Quantity = 1
+}
+case object RecipientsType extends EntityType {
+  override def asString(): String = "recipients"
+
+  override def extractQuantity(mail: Mail): Quantity = Quantity.liftOrThrow(mail.getRecipients.size())
+}
+case object Size extends EntityType {
+  override def asString(): String = "size"
+
+  override def extractQuantity(mail: Mail): Quantity = Quantity.liftOrThrow(mail.getMessageSize.toInt)
+}
+case object TotalSize extends EntityType {
+  override def asString(): String = "totalSize"
+
+  override def extractQuantity(mail: Mail): Quantity = Quantity.liftOrThrow((mail.getMessageSize * mail.getRecipients.size()).toInt)
+}
+
+case class PerSenderRateLimiter(rateLimiter: Option[RateLimiter], keyPrefix: Option[KeyPrefix], entityType: EntityType) {
+  def rateLimit(sender: MailAddress, mail: Mail): Publisher[RateLimitingResult] = {
+    val rateLimitingKey = SenderKey(keyPrefix, entityType, sender)
+
+    rateLimiter.map(limiter => limiter.rateLimit(rateLimitingKey, entityType.extractQuantity(mail)))
+      .getOrElse(SMono.just[RateLimitingResult](AcceptableRate))
+  }
+}
+
+case class SenderKey(keyPrefix: Option[KeyPrefix], entityType: EntityType, mailAddress: MailAddress) extends RateLimitingKey {
+  override def asString(): String = keyPrefix.map(prefix => s"${prefix.value}_${entityType.asString()}_${mailAddress.asString()}")
+    .getOrElse(s"${entityType.asString()}_${mailAddress.asString()}")
+}
+
+/**
+ * <p><b>PerSenderRateLimit</b> allows defining and enforcing rate limits for the sender of matching emails.</p>
+ *
+ * <ul>This allows writing rules like:
+ *   <li>A sender can send 10 emails per hour</li>
+ *   <li>A sender can send email to a total of 20 recipients per hour</li>
```

Comment:
I don't remember the number of recipients as a rate limit being discussed? Is this new?

Reply:
Yes this is new. While writing the code I just elt like this ade sense. This was easy to add, so I did.

Retrospectively this is required to implement some limits defined eg on the fastmail limit page (https://www.fastmail.help/hc/en-us/articles/1500000277382-Account-limits) which we uses as an inspiration for this feature.

Reply:
Well I'm ok with it, maybe should add it quickly on the ticket though :)